### ISSUE TITLE
T.must autocorrect doesn't operate on constructor kwargs

### ISSUE THREAD

**Issue by bhuga** [sorbet/sorbet#3498]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(package%3A%20Integer).void%7D%0A%20%20def%20initialize(package)%0A%20%20end%0A%0A%20%20def%20self.foo%0A%20%20%20%20pkg%20%3D%20T.let(nil%2C%20T.nilable(Integer))%0A%0A%20%20%20%20A.new(package%3A%20pkg)%0A%20%20%20%205.coerce(pkg)%20%20%0A%20%20%20%20end%0Aend%0A%0A)

```ruby
# typed: true
class A
    extend T::Sig

  sig {params(package: Integer).void}
  def initialize(package)
  end

  def self.foo
    pkg = T.let(nil, T.nilable(Integer))

    A.new(package: pkg)
    5.coerce(pkg)  
    end
end


```

#### Observed output

```
editor.rb:12: Expected Integer but found {package: T.nilable(Integer)} for argument package https://srb.help/7002
    12 |    A.new(package: pkg)
                  ^^^^^^^^^^^^
    editor.rb:5: Method A#initialize has specified package as Integer
     5 |  sig {params(package: Integer).void}
                      ^^^^^^^
  Got {package: T.nilable(Integer)} originating from:
    editor.rb:12:
    12 |    A.new(package: pkg)
                  ^^^^^^^^^^^^

editor.rb:13: Expected Numeric but found T.nilable(Integer) for argument arg0 https://srb.help/7002
    13 |    5.coerce(pkg)  
            ^^^^^^^^^^^^^
    https://github.com/sorbet/sorbet/tree/master/rbi/core/integer.rbi#L643: Method Integer#coerce has specified arg0 as Numeric
     643 |        arg0: Numeric,
                  ^^^^
  Got T.nilable(Integer) originating from:
    editor.rb:10:
    10 |    pkg = T.let(nil, T.nilable(Integer))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Autocorrect: Use `-a` to autocorrect
    editor.rb:13: Replace with T.must(pkg)
    13 |    5.coerce(pkg)  
                     ^^^
Errors: 2
```

#### Expected behavior

Only one incorrect nilable is autocorrect. T.must should be equally valid in both cases.

- - -

**Comment by jez** [comment]
Relates to #1474

**Comment by froydnj** [comment]
I looked at this a little this morning.  There are a couple different issues here:

1. We have overly-broad location information for keyword arguments, which I filed as #4387 
2. Related to the above, we pass in `Loc::none()` for matching keyword argument types:

https://github.com/sorbet/sorbet/blob/16403b337aa621a15ae731e0c6d367926033560d/core/types/calls.cc#L997-L1005

which causes the type-driven autocorrect in `matchArgType`:

https://github.com/sorbet/sorbet/blob/16403b337aa621a15ae731e0c6d367926033560d/core/types/calls.cc#L188-L200

to not kick in:

https://github.com/sorbet/sorbet/blob/16403b337aa621a15ae731e0c6d367926033560d/core/TypeDrivenAutocorrect.cc#L7-L11

as `Loc::none()` is `!exists()`. Fixing #4387 will automagically add the autocorrect.

3. But the third issue is that in the original testcase, we never actually hit that code, because `initialize` takes a positional arg and the `A.new` call uses a keyword arg, so we go through:

https://github.com/sorbet/sorbet/blob/16403b337aa621a15ae731e0c6d367926033560d/core/types/calls.cc#L872-L885

Notice the shape type in the error message:

```
  Got {package: T.nilable(Integer)} originating from:
    editor.rb:12:
    12 |    A.new(package: pkg)
                  ^^^^^^^^^^^^
```

So I think the right answer for _this_ particular bug is to...maybe detect that the user is passing keyword arguments but the method actually takes positional ones?  I don't think we could offer an autocorrect in that case.

**Comment by jez** [comment]
On further inspection, it looks like there was a typo on line 6 (`package` -> `package:`) which would be required for the original request in the bug report to make sense. With that typo fix, @froydnj 's change in #4400 actually fixes this, so we can close! 🎉
